include/graphblas/reference/blas3.hpp

```diff
@@ -916,6 +916,206 @@ namespace grb {
+			"called with an output matrix that does not match the output domain of "
+			"the given multiplication operator" );
+		static_assert( ( !(descr & descriptors::invert_mask)
+			),
+			"grb::maskedOuter: invert_mask descriptor cannot be used ");
+#ifdef _DEBUG
+		std::cout << "In grb::maskedOuter (reference)\n";
+#endif
+
+		const size_t nrows = size( u );
+		const size_t ncols = size( v );
+
+		const size_t m = grb::nrows( mask );
+		const size_t n = grb::ncols( mask );
+
+		assert( phase != TRY );
+		if( nrows != grb::nrows( A ) || nrows != m ) {
+			return MISMATCH;
+		}
+
+		if( ncols != grb::ncols( A ) || ncols != n ) {
+			return MISMATCH;
+		}
+
+
+		const auto &mask_raw = internal::getCRS( mask );
+
+		size_t nzc = 0;
+		for( size_t i = 0; i < nrows; ++i ) {
+			if( internal::getCoordinates( u ).assigned( i ) ) {
+				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
+					const size_t k_col = mask_raw.row_index[ k ];
+					if( internal::getCoordinates( v ).assigned( k_col ) ) {
+						nzc++;
+					}
+				}
+			}
+		}
+
+		if( phase == RESIZE ) {
+			return resize( A, nzc );
+		}
+
+		assert( phase == EXECUTE );
+		if( capacity( A ) < nzc ) {
+#ifdef _DEBUG
+			std::cout << "\t insufficient capacity to complete "
+				"requested masked outer-product computation\n";
+#endif
+			const RC clear_rc = clear( A );
+			if( clear_rc != SUCCESS ) {
+				return PANIC;
+			} else {
+				return FAILED;
+			}
+		}
+
+		grb::Monoid<
+			grb::operators::left_assign< OutputType >,
+			grb::identities::zero
+		> monoid;
+
+		RC ret = SUCCESS;
+		if( phase == EXECUTE ) {
+			ret = grb::clear( A );
+		}
+		assert( nnz( A ) == 0 );
+
+		auto &CRS_raw = internal::getCRS( A );
+		auto &CCS_raw = internal::getCCS( A );
+
+		const InputType1 * __restrict__ const x = internal::getRaw( u );
+		const InputType2 * __restrict__ const y = internal::getRaw( v );
+		char * arr = nullptr;
+		char * buf = nullptr;
+		OutputType * valbuf = nullptr;
+		internal::getMatrixBuffers( arr, buf, valbuf, 1, A );
+		config::NonzeroIndexType * A_col_index = internal::template
+			getReferenceBuffer< typename config::NonzeroIndexType >( ncols + 1 );
+
+		
+		internal::Coordinates< reference > coors;
+		coors.set( arr, false, buf, ncols );
+
+		CRS_raw.col_start[ 0 ] = 0;
+		for( size_t j = 0; j <= ncols; ++j ) {
+			CCS_raw.col_start[ j ] = 0;
+		}
+
+		
+		nzc = 0;
+		for( size_t i = 0; i < nrows; ++i ) {
+			if( internal::getCoordinates( u ).assigned( i ) ) {
+				for( auto k = mask_raw.col_start[ i ]; k < mask_raw.col_start[ i + 1 ]; ++k ) {
+					const size_t k_col = mask_raw.row_index[ k ];
+					if( internal::getCoordinates( v ).assigned( k_col ) ) {
+						(void) ++nzc;
+						(void) ++CCS_raw.col_start[ k_col + 1 ];
+					}
+				}
+			}
+			CRS_raw.col_start[ i + 1 ] = nzc;
+		}
+
+
+		A_col_index[ 0 ] = 0;
+		for( size_t j = 1; j < ncols; ++j ) {
+			CCS_raw.col_start[ j + 1 ] += CCS_raw.col_start[ j ];
+			A_col_index[ j ] = 0;
+		}
```

Comment:
Split this loop in two and assign `A_col_index` in a separate `#pragma omp parallel for simd`

Reply:
Done, thanks!